core/consensus/babe/primitives/src/lib.rs

```diff
@@ -0,0 +1,42 @@
+// Copyright 2017-2019 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Primitives for BABE.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use runtime_primitives::ConsensusEngineId;
+use substrate_client::decl_runtime_apis;
+
+/// The `ConsensusEngineId` of BABE.
+pub const BABE_ENGINE_ID: ConsensusEngineId = [b'b', b'a', b'b', b'e'];
+
+decl_runtime_apis! {
+	/// API necessary for block authorship with BABE.
+	pub trait BabeApi {
+		/// Return the slot duration in milliseconds for BABE.  Currently, only
+		/// the value provided by this type at genesis will be used.
+		///
+		/// Dynamic slot duration may be supported in the future.
+		fn slot_duration() -> u64;
+
+		/// Return the expected block time in milliseconds for BABE.  Currently,
+		/// only the value provided by this type at genesis will be used.
+		///
+		/// Dynamic expected block time may be supported in the future.
+		fn expected_block_time() -> u64;
```

Comment:
Perhaps combine this with the other function?

Reply:
Should it return a struct?

Reply:
That seems slightly more heavyweight.

Reply:
Or a tuple

Reply:
@rphmeier I am not a fan of a tuple here, because it could easily lead to confusion (which `u64` means what?).  Is there a drawback of having two separate functions?  For instance, will these functions be called indirectly via a trait object vtable?

Reply:
These two functions are only really called at genesis for initializing the engine, but for consistency's sake, it might help to know how it would fit in with swappable consensus.

my rough vision for swappable consensus is that there will be a special consensus engine with `OVERSEER_ENGINE_ID`. A runtime will send out a `DigestItem::Consensus(OVERSEER_ENGINE_ID, Signal::Swap(SOME_OTHER_ENGINE, engine_startup_data).encode())`

The `startup_data` for BABE will be a struct containing both members, and current timestamp, so we'll need the struct anyway. Although at genesis we just assume timestamp=0.

Reply:
@rphmeier Sadly, using a struct breaks the macro I was using.  Should I ditch the macro entirely?  I think that is the best option.

Reply:
You can not ditch the macro. How does a struct as return argument break the macro?

The macro needs to be defined outside of the macro.

Reply:
The macro is gone, and a custom struct is used.